crypto/random/rand_test.go

```diff
@@ -1,263 +1,405 @@
+	//  - permuting an empty set returns an empty list
+	//  - drawing a sample of size zero from a non-empty set returns an empty list
+	t.Run("empty sets", func(t *testing.T) {
+
+		// verify that permuting an empty set returns an empty list
+		res, err := rng.SubPermutation(0, 0)
+		require.NoError(t, err)
+		assert.True(t, len(res) == 0)
+
+		// verify that drawing a sample of size zero from a non-empty set returns an empty list
+		res, err = rng.SubPermutation(10, 0)
+		require.NoError(t, err)
+		assert.True(t, len(res) == 0)
+	})
+
+	t.Run("negative inputs", func(t *testing.T) {
+		res, err := rng.Permutation(-3)
+		require.Error(t, err)
+		assert.Nil(t, res)
+
+		res, err = rng.SubPermutation(5, -3)
+		require.Error(t, err)
+		assert.Nil(t, res)
+
+		res, err = rng.SubPermutation(-3, 5)
+		require.Error(t, err)
+		assert.Nil(t, res)
 	})
-	require.NoError(t, err)
-	assert.True(t, len(emptySlice) == 0)
 }
 
-func TestRandomSamples(t *testing.T) {
-	listSize := 100
-	samplesSize := 20
-	seed := make([]byte, 16)
-	seed[0] = 45
-	rng, err := NewRand(seed)
+// Simple unit testing of Shuffle using a very basic randomness test.
+// It doesn't evaluate randomness of the output and doesn't perform advanced statistical tests.
+func TestShuffle(t *testing.T) {
+
+	seed := make([]byte, Chacha20SeedLen)
+	crand.Read(seed)
+	customizer := make([]byte, Chacha20CustomizerMaxLen)
+	crand.Read(customizer)
+
+	rng, err := NewChacha20PRG(seed, customizer)
 	require.NoError(t, err)
-	// statictics parameters
-	sampleSize := 100000
-	tolerance := 0.05
-	// tests the subset sampling randomness
-	samplingDistribution := make([]float64, listSize)
-	// tests the subset ordering randomness (using a particular element testElement)
-	orderingDistribution := make([]float64, samplesSize)
-	testElement := rand.Intn(listSize)
-	// Slice to shuffle
-	list := make([]int, 0, listSize)
-	for i := 0; i < listSize; i++ {
-		list = append(list, i)
-	}
 
-	for i := 0; i < sampleSize; i++ {
-		err = rng.Samples(listSize, samplesSize, func(i, j int) {
-			list[i], list[j] = list[j], list[i]
-		})
-		require.NoError(t, err)
-		has := make(map[int]struct{})
-		for j, e := range list[:samplesSize] {
-			// check for repetition
-			if _, ok := has[e]; ok {
-				t.Errorf("dupplicated item in the results returned by PermutateSubset")
-			}
-			has[e] = struct{}{}
-			// fill the distribution
-			samplingDistribution[e] += 1.0
-			if e == testElement {
-				orderingDistribution[j] += 1.0
+	t.Run("basic randmoness", func(t *testing.T) {
+		listSize := 100
+		// test parameters
+		sampleSize := 100000
+		tolerance := 0.05
+		// the distribution of a particular element of the list, testElement
+		distribution := make([]float64, listSize)
+		testElement := mrand.Intn(listSize)
+		// Slice to shuffle
+		list := make([]int, 0, listSize)
+		for i := 0; i < listSize; i++ {
+			list = append(list, i)
+		}
+
+		for i := 0; i < sampleSize; i++ {
+			err = rng.Shuffle(listSize, func(i, j int) {
```

Comment:
The test here run for 100000 times, and each time it will shuffle the list, and pick an item at the same index, to check if the picked 100000 items are distributed evenly across all elements in the list.

My question is why do we reuse the shuffled list and shuffle again on top of the shuffled list, rather than reset the list to start from `1` to `100`, and shuffle again. 

Technically, when user uses the shuffle API, they would only call Shuffle once, rather than shuffle the list for X times. So they would expect the shuffled list is random after one time shuffling. So I think the test should also test the property under the same scenario. 

It also applies to the testing to other methods, such as `Permutation`, `SubPermutation`

Reply:
Both experiments (the implemented one or the suggested one) should give the same result if the outcome is uniformly distributed. The experiment should be: 
 1. start from any permutation of the array
 2. shuffle using the shuffle to test
 3. check what's in the some constant index
 What you're suggesting is to always start from the same specific permutation in step 1 (which can be [1,2,3..,100]). While what I implemented is to start from a different permutation each time. Instead of choosing a random permutation, I just re-use the latest permutation as an optimization. 
 Both ways must give the a uniform result for the implementation to be valid. I can add another the suggested way too.

Reply:
added in https://github.com/onflow/flow-go/pull/1782/commits/60d5d19df1351645725a780c4c1f4c1ee536f17b